torch/optim/asgd.py

```diff
@@ -318,25 +332,41 @@ def _view_complex_as_real(tensor_list):
             else:
                 grouped_grads = torch._foreach_add(grouped_grads, grouped_params, alpha=weight_decay)
 
-        # decay term
-        eta = _get_value(grouped_etas[0])
-        torch._foreach_mul_(grouped_params, 1 - lambd * eta)
-
-        # update parameter
-        torch._foreach_add_(grouped_params, grouped_grads, alpha=-eta)
-
-        # averaging
-        for i in range(len(grouped_axs)):
-            if is_compiling() or grouped_mus[i].item() != 1:
-                grouped_axs[i].add_(grouped_params[i].sub(grouped_axs[i]).mul(grouped_mus[i]))
-            else:
-                grouped_axs[i].copy_(grouped_params[i])
-
-        # update eta and mu
-        for i in range(len(grouped_mus)):
-            new_eta = _to_tensor(
-                lr / (1 + lambd * lr * _get_value(grouped_state_steps[i]) ** alpha)
-            )
-            grouped_etas[i].copy_(new_eta)
-            new_mu = _to_tensor(1 / max(1, _get_value(grouped_state_steps[i]) - t0))
-            grouped_mus[i].copy_(new_mu)
+        if capturable:
+            # decay term
+            decay = torch._foreach_add(torch._foreach_mul(etas, -lambd), 1)
+            # update parameter
+            torch._foreach_mul_(grouped_params,
+                                torch._foreach_add(torch._foreach_div(torch._foreach_mul(grouped_grads,
+                                                                                         torch._foreach_mul(etas, -1)),
+                                                                      torch._foreach_mul(grouped_params, decay)),
+                                                   1.0))
+
+            torch._foreach_add_(grouped_axs, torch._foreach_mul(torch._foreach_sub(grouped_params, grouped_axs), grouped_mus))
+            torch._foreach_copy_(grouped_mus,
+                                 torch._foreach_reciprocal(torch._foreach_maximum(torch._foreach_sub(grouped_state_steps, t0),
+                                                                                  1.0)))
+        else:
+            # decay term
+            eta = _get_value(grouped_etas[0])
+            torch._foreach_mul_(grouped_params, 1 - lambd * eta)
+
+            # update parameter
+            torch._foreach_add_(grouped_params, grouped_grads, alpha=-eta)
+
+
+            # averaging
+            for i in range(len(grouped_axs)):
+                if is_compiling() or grouped_mus[i].item() != 1:
+                    grouped_axs[i].add_(grouped_params[i].sub(grouped_axs[i]).mul(grouped_mus[i]))
+                else:
+                    grouped_axs[i].copy_(grouped_params[i])
```

Comment:
We should just rewrite this entire portion with a lerp, and this part should also be shared with the capturable impl.

```suggestion
            # averaging: ax = ax + mu * (param - ax)
            torch._foreach_lerp_(grouped_axs, grouped_params, grouped_mus)
```